components/raftstore/src/store/fsm/apply.rs

```diff
@@ -1631,134 +1744,90 @@ where
+                                        "ignore remove unmatched peer";
+                                        "region_id" => self.region_id(),
+                                        "peer_id" => self.id(),
+                                        "expect_peer" => ?peer,
+                                        "get_peeer" => ?p
+                                    );
+                                    return Err(box_err!(
+                                        "remove unmatched peer: expect: {:?}, get {:?}, ignore",
+                                        peer,
+                                        p
+                                    ));
+                                }
+                                if self.id == peer.get_id() {
+                                    // Remove ourself, we will destroy all region data later.
+                                    // So we need not to apply following logs.
+                                    self.stopped = true;
+                                    self.pending_remove = true;
+                                }
+                            }
+                            None => unreachable!(),
+                        }
+                    }
+                },
             }
+            confchange_cmd_metric::inc_success(change_type);
+            info!(
+                "conf change successfully";
+                "type" => ?change_type,
+                "region_id" => self.region_id(),
+                "peer_id" => self.id(),
+                "peer" => ?peer,
+                "region" => ?&self.region
+            );
         }
+        let conf_ver = region.get_region_epoch().get_conf_ver() + changes.len() as u64;
+        region.mut_region_epoch().set_conf_ver(conf_ver);
+        Ok(region)
+    }
 
-        let state = if self.pending_remove {
-            PeerState::Tombstone
-        } else {
-            PeerState::Normal
-        };
-        if let Err(e) = write_peer_state(ctx.kv_wb_mut(), &region, state, None) {
-            panic!("{} failed to update region state: {:?}", self.tag, e);
+    fn leave_joint(&self) -> Result<Region> {
+        let mut region = self.region.clone();
+        for peer in region.mut_peers().iter_mut() {
+            match peer.get_role() {
+                PeerRole::IncomingVoter => peer.set_role(PeerRole::Voter),
+                PeerRole::DemotingVoter => peer.set_role(PeerRole::Learner),
+                _ => {}
+            }
         }
-
-        let mut resp = AdminResponse::default();
-        resp.mut_change_peer().set_region(region.clone());
-
-        Ok((
-            resp,
-            ApplyResult::Res(ExecResult::ChangePeer(ChangePeer {
-                index: ctx.exec_ctx.as_ref().unwrap().index,
-                conf_change: Default::default(),
-                peer: peer.clone(),
-                region,
-            })),
-        ))
+        // TODO: Should we also increase epoch here?
```

Comment:
I think so. The region has been changed, client need epoch change to decide whether update local cache.